File: ocr_reader.cpp
// ocr_reader.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <tesseract/baseapi.h>
#include <leptonica/allheaders.h>

using namespace std;

string extractText(const string& imagePath, const string& lang) {
    tesseract::TessBaseAPI* api = new tesseract::TessBaseAPI();
    if (api->Init(NULL, lang.c_str())) {
        cerr << "Could not initialize tesseract.\n";
        delete api;
        return "";
    }
    Pix* image = pixRead(imagePath.c_str());
    if (!image) {
        cerr << "Could not read image: " << imagePath << "\n";
        delete api;
        return "";
    }
    api->SetImage(image);
    char* outText = api->GetUTF8Text();
    string result = outText ? string(outText) : "";
    delete[] outText;
    pixDestroy(&image);
    api->End();
    delete api;
    return result;
}

string exportDocx(const string& text, const string& outputPath) {
    // Simple DOCX export using a minimal template
    ofstream f(outputPath);
    f << "DOCX export requires a library. Using TXT fallback.\n";
    f << text;
    f.close();
    return outputPath;
}

string exportTxt(const string& text, const string& outputPath) {
    ofstream f(outputPath);
    f << text;
    f.close();
    return outputPath;
}

int main(int argc, char* argv[]) {
    vector<string> images;
    string output = "output.docx";
    string lang = "eng";
    string format = "docx";
    int opt;
    while ((opt = getopt(argc, argv, "o:l:f:")) != -1) {
        switch (opt) {
            case 'o': output = optarg; break;
            case 'l': lang = optarg; break;
            case 'f': format = optarg; break;
            default: break;
        }
    }
    for (int i = optind; i < argc; i++) {
        images.push_back(argv[i]);
    }
    if (images.empty()) {
        cerr << "Error: No images provided.\n";
        return 1;
    }
    cout << "📄 OCR Reader\n";
    cout << "Images: ";
    for (size_t i = 0; i < images.size(); i++) {
        if (i) cout << ", ";
        cout << images[i];
    }
    cout << "\nLanguage: " << lang << "\n";
    vector<string> allText;
    for (const auto& imgPath : images) {
        cout << "Processing: " << imgPath << "\n";
        string text = extractText(imgPath, lang);
        if (!text.empty()) {
            allText.push_back("--- " + imgPath + " ---\n" + text);
        } else {
            cout << "⚠️ No text extracted from " << imgPath << "\n";
        }
    }
    if (allText.empty()) {
        cerr << "❌ No text extracted from any image.\n";
        return 1;
    }
    string combinedText;
    for (const auto& t : allText) {
        combinedText += t + "\n";
    }
    cout << "\n📖 Extracted text:\n";
    if (combinedText.length() > 500) {
        cout << combinedText.substr(0, 500) << "...\n";
    } else {
        cout << combinedText;
    }
    string outputPath;
    if (format == "docx") {
        outputPath = exportDocx(combinedText, output);
        cout << "\n📄 Exporting to DOCX... ✅\n";
    } else {
        outputPath = exportTxt(combinedText, output);
        cout << "\n📄 Exporting to TXT... ✅\n";
    }
    cout << "File saved: " << outputPath << "\n";
    return 0;
}

File: ocr_reader.php
# ocr_reader.php
#!/usr/bin/env php
<?php

require_once 'vendor/autoload.php';

use thiagoalessio\TesseractOCR\TesseractOCR;
use PhpOffice\PhpWord\PhpWord;
use PhpOffice\PhpWord\IOFactory;

function extractText($imagePath, $lang) {
    try {
        $text = (new TesseractOCR($imagePath))
            ->lang($lang)
            ->run();
        return trim($text);
    } catch (Exception $e) {
        fwrite(STDERR, "Error processing $imagePath: " . $e->getMessage() . "\n");
        return null;
    }
}

function exportDocx($text, $outputPath) {
    $phpWord = new PhpWord();
    $section = $phpWord->addSection();
    $section->addTitle('OCR Extracted Text', 1);
    $section->addText($text);
    $writer = IOFactory::createWriter($phpWord, 'Word2007');
    $writer->save($outputPath);
    return $outputPath;
}

function exportTxt($text, $outputPath) {
    file_put_contents($outputPath, $text);
    return $outputPath;
}

$opts = getopt("", ["output:", "lang:", "format:"]);
$images = array_slice($argv, 1);
foreach ($images as $i => $img) {
    if (strpos($img, '--') === 0) {
        unset($images[$i]);
    }
}
$images = array_values($images);

if (empty($images)) {
    fwrite(STDERR, "Error: No images provided.\n");
    exit(1);
}

$output = $opts['output'] ?? 'output.docx';
$lang = $opts['lang'] ?? 'eng';
$format = $opts['format'] ?? 'docx';

echo "📄 OCR Reader\n";
echo "Images: " . implode(', ', $images) . "\n";
echo "Language: $lang\n";

$allText = [];
foreach ($images as $imgPath) {
    echo "Processing: $imgPath\n";
    $text = extractText($imgPath, $lang);
    if ($text) {
        $allText[] = "--- $imgPath ---\n$text";
    } else {
        echo "⚠️ No text extracted from $imgPath\n";
    }
}

if (empty($allText)) {
    fwrite(STDERR, "❌ No text extracted from any image.\n");
    exit(1);
}

$combinedText = implode("\n", $allText);
echo "\n📖 Extracted text:\n";
if (strlen($combinedText) > 500) {
    echo substr($combinedText, 0, 500) . "...\n";
} else {
    echo $combinedText . "\n";
}

if ($format === 'docx') {
    $outputPath = exportDocx($combinedText, $output);
    echo "\n📄 Exporting to DOCX... ✅\n";
} else {
    $outputPath = exportTxt($combinedText, $output);
    echo "\n📄 Exporting to TXT... ✅\n";
}
echo "File saved: $outputPath\n";
?>
